gpui: Add support for touch events on Wayland

robert7k opened the pull request:
This PR works towards #13698 and implements touch events on Wayland. I'm aware of #34675, but that PR has probably been rebased incorrectly, so I couldn't find the source code or get it to work.

Currently, zed can't be used on a touch screen on Wayland: the touch events are simply ignored, nothing happens when finger-clicking anywhere in zed. This PR makes zed usable on a touch-screen.

I have tested this on Linux with Wayland.

The PR contains multiple features:

- process touch evens to allow clicking on UI elements
- allows scrolling using a two-finger gesture
- pinch to zoom gesture

Release Notes:

- Added support for touch events on Wayland


Demo video (the red and blue circles show the touch events. Unfortunately, not all touch events are properly visible, but everything in the video has been done through a touch screen):


https://github.com/user-attachments/assets/eed3f04c-1c14-41a2-9b89-c1cee2ed82f5

robert7k committed:
Process touch events

robert7k committed:
Add two-finger scroll support

robert7k committed:
Add pinch-to-zoom gesture support

robert7k committed:
Format code

ecstatic-morse commented on crates/gpui/src/platform/linux/wayland/client.rs:
```diff
@@ -2157,3 +2198,332 @@ impl Dispatch<zwp_primary_selection_source_v1::ZwpPrimarySelectionSourceV1, ()>
+                            state.button_pressed = Some(MouseButton::Left);
+                            drop(state);
+                            window.handle_input(input);
+                        }
```
Seems like `state.mouse_focused_window` is always `None`, so `MouseUp`/`ScrollWheel`  are never sent on my machine.

robert7k committed:
Merge branch 'main' into feature/touch-events

robert7k commented:
@MrSubidubi thank you for getting back to this PR.

I merged the latest main and resolved the merge conflicts. The feature is now working again. I'd be happy to push this forward to finally land this feature!

robert7k committed:
Fix formatting

robert7k committed:
Update touch scroll sensitivity and tracking logic

Adjust touch scroll sensitivity and track last scroll
center for more responsive scrolling

robert7k committed:
Merge branch 'main' into feature/touch-events

GitGlimpse895 commented:
Hey @yara-blue — just checking in on this one. @robert7k rebased and fixed merge conflicts last month and has confirmed it's working again. The original `mouse_focused_window` bug flagged by @ecstatic-morse back in November — has that been addressed in the latest commits?

I can confirm that current main only has pinch gesture support via `zwp_pointer_gesture_pinch_v1` (pointer gestures protocol), but actual finger touch via `wl_touch` is still missing entirely. This PR covers that gap. Would love to see a review pass so this can finally land!

ryanabx commented:
Maybe this can be revisited when https://github.com/zed-industries/zed/pull/63068 lands? I assume at that point GPUI will have more touch input abstractions in place.
